Bug 1177597 & 1183210 - Search screen does not show on iOS9

codestergit opened the pull request:
In iOS9 due to some bug `UIKeyInput` method `insertText` is not calling.
This commit resolve the issue.

thebnich commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -70,13 +76,13 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
         }
-
-        return true
+        removeCompletion()
```
I don't think we want to remove the completion every time a character is entered. This causes the text to flicker every time a character is entered.

thebnich commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -70,13 +76,13 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
     }
 
     func textField(textField: UITextField, shouldChangeCharactersInRange range: NSRange, replacementString string: String) -> Bool {
+        var allowChractersChange = true
```
Nit: `Chracters` -> `Characters`

thebnich commented:
Can you explain these changes, maybe adding some comments? I understand the `insertText` -> `SELtextDidChange` if `insertText` isn't working, but I'm not clear on what all the other changes are doing.

codestergit commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -70,13 +76,14 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
-
-        return true
+        //Removes if any autocompletion before changing the chracters.
+        removeCompletion()
```
Removes Autocompletion before changing any characters.

codestergit commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -122,13 +129,21 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
+        if characterDeleted {
+            characterDeleted = false
+            
+            return
```
If any character is deleted than we simply returns. We do not want autocompletion.

codestergit commented:
@thebnich  added comments. Please let me know if any further clarification needed.

thebnich commented:
This seems to work well on English keyboards, though I found a regression switching to Japanese.

Steps:
 1. Switch to Japanese keyboard.
 2. Click URL bar.
 3. Type "ha".

This should show the Hiragana symbol for "ha" (**あ**); however, it shows **hあ**, so it looks like this change is breaking text composition somehow. I filed [bug 1183914](https://bugzilla.mozilla.org/show_bug.cgi?id=1183914) to get some automated tests for non-English text entry -- it's very fragile!

fxios-buildasaur commented:
Result of integration 1
Integration took 43 seconds.
? build errors: checkout-error

codestergit commented:
@thebnich updated the pull request. Fixed flicker issue and Japanese character issue.

codestergit commented:
Hi @thebnich,
Resolved conflicts. Now can be merged.

codestergit commented:
Hi @thebnich  It also resolve [Bug-1183210] ( https://bugzilla.mozilla.org/show_bug.cgi?id=1183210) flicker issue while entering text.

codestergit commented:
@thebnich  removed `debounce` added for [Bug 1173162 ] (https://bugzilla.mozilla.org/show_bug.cgi?id=1173162). In this PR we are hitting db when required i.e if user enter text does not match with saved suggestion.

codestergit commented:
@thebnich  updated the pull request. Resolved conflicts.

codestergit commented:
@thebnich  updated the pull request addressed comment. I have made separate commit for that and squashed in to "Hitting history db only when required". I hope that is fine.

codestergit commented:
Hi @thebnich 
Due to lag between `shouldChangeCharactersInRange ` and `textDidChange` event we need to do this. detailed [comment] (https://gist.github.com/thebnich/c0e5731f9e25b24b1168) is there. I will try to find some alternative. I need your thoughts further. Thanks for all your help :)

codestergit commented:
@thebnich 
Updated the pull request. Thanks for your suggestion. I have simplified the patch.
Thanks

codestergit commented:
Hi @thebnich 
Updated pull request and resolved conflicts.
During texting I have found two issues which is regression of [Bug-1166781](https://bugzilla.mozilla.org/show_bug.cgi?id=1166781).
I have done separate pull request for them also.
https://github.com/mozilla/firefox-ios/pull/807
https://github.com/mozilla/firefox-ios/pull/808

thebnich commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -39,9 +40,11 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
             if self.editing {
-                self.autocompleteDelegate?.autocompleteTextField(self, didEnterText: self.text)
+                let enteredText = self.text.substringToIndex(advance(self.text.startIndex, self.enteredTextLength, self.text.endIndex))
+                self.autocompleteDelegate?.autocompleteTextField(self, didEnterText: enteredText)
```
Do we need this change? I think it's safe to just use `self.text` here.

codestergit replied:
@thebnich  Here we need `enteredText` because it is also called from `SELtextDidChange`. Here we are having complete string with completion not the user entered text.




Updated the pull request.

thebnich replied:
OK. I see we are doing this `enteredText` calculation in lots of places. Maybe it would make sense to just store `enteredText` directly as a class variable, and remove `enteredTextLength`? If there are places we still need the entered text length, we could just use `count(enteredText)`. And then we don't need to use `substringToIndex` everywhere.

codestergit replied:
@thebnich  You always know what is best :).
 Changed `enteredTextLength ` to `entredText` working good.
As we are not calculating `enteredTextLength` when in completion mode and if user enters some text in between than it will be calculated in `SELtextDidChange `. Thanks for your suggestion.

thebnich commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -82,27 +87,51 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
         return true
     }
 
+    private func  removeCompletionIfRequiredForEnteredString(string: String) {
```
Nit: remove extra space after `func`

thebnich commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -82,27 +87,51 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
 
+    private func  removeCompletionIfRequiredForEnteredString(string: String) {
+        let actualEnteredString = text.substringToIndex(advance(text.startIndex, enteredTextLength, text.endIndex)) + string
+        // If actual entred text by user not starts with previous suggestion than remove completion.
```
Nit: `If user-entered text does not start with previous suggestion then remove the completion.`

thebnich commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -82,27 +87,51 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
-        if let suggestion = suggestion where editing {
+        // Setting auto complete suggestion in between breaks input sesssion. For multiple input characters 
+        // while text is marked than user is not entered its full input. So here we are checking `markedTextRange`. 
+        // If it is nil it means now multiinput character session is completed and these character can not change.
```
Some adjustments:
```
// Setting the autocomplete suggestion during multi-stage input will break the session since the text
// is not fully entered. If `markedTextRange` is nil, that means the multi-stage input is complete, so
// it's safe to append the suggestion.
```

thebnich commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -82,27 +87,51 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
+        // Setting auto complete suggestion in between breaks input sesssion. For multiple input characters 
+        // while text is marked than user is not entered its full input. So here we are checking `markedTextRange`. 
+        // If it is nil it means now multiinput character session is completed and these character can not change.
+        if let suggestion = suggestion where editing && markedTextRange == nil{
```
Nit: `nil{` -> `nil {`

thebnich commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -82,27 +87,51 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
                 completionActive = true
+                previousSuggestion = suggestion
+            }
+            else if completionActive && enteredText == suggestion {
```
Nit: `} else if ... {` on same line

thebnich commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -86,27 +91,50 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
                 completedAndMarkedString.addAttribute(NSBackgroundColorAttributeName, value: AutocompleteTextFieldUX.HighlightColor, range: NSMakeRange(enteredTextLength, count(endingString)))
                 attributedText = completedAndMarkedString
                 completionActive = true
+                previousSuggestion = suggestion
```
How about we return early here and remove the `if else` below? Then, at the very bottom of this function outside of all the `if` statements, you do `removeCompletion`? That way, if the setting the suggestion fails for any reason, we always clear the existing completion if there is one.

codestergit replied:
Thanks great :+1:. 
Done that

thebnich commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -136,12 +164,16 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
+            // suggestion and if it is valid than we are showing the completion.
+            setAutocompleteSuggestion(previousSuggestion)
+        }
+        else {
```
Nit: `} else {`

thebnich commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -136,12 +164,16 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
-
+        if completionActive {
+            // As suggestion may not be availaible immediate so we are caching previous
+            // suggestion and if it is valid than we are showing the completion.
```
I'd just reduce this to: `// Immediately reuse the previous suggestion if it's still valid.`

thebnich commented on Client/Frontend/Widgets/AutocompleteTextField.swift:
```diff
@@ -150,7 +182,6 @@ class AutocompleteTextField: UITextField, UITextFieldDelegate {
         removeCompletion()
         super.deleteBackward()
         enteredTextLength = count(self.text)
-        autocompleteDelegate?.autocompleteTextField(self, didEnterText: self.text)
```
I'm not sure we want to remove this line -- this is what triggers the search results below the text field. That is, if I type "y", I should see all my history entries below that contain "y". After [bug 1166781](https://bugzilla.mozilla.org/show_bug.cgi?id=1166781) this line should be safe to keep hopefully?

codestergit replied:
@thebnich I think we can safely remove this line. On master branch we are using `insertText` which calls only if we are entering text not deleting it. Here we are using `textChange` event and there we are notifying all change. However that will call with `debounce` delay. I do not think that is a problem.

codestergit commented:
@thebnich  Thanks for suggestion. Updated the pull request.
Please let me know if any other changes required.
